intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+# --------------------------------
+# When a module is parametrized, we find that the module has changed a bit.
+# We may observe this by simply printing the module
+layer = nn.Linear(3, 3)
+print(f"Unparametrized:\n{layer}")
+P.register_parametrization(layer, "weight", Symmetric())
+print(f"Parametrized:\n{layer}")
+
+###############################################################################
+# We see that the ``Symmetric`` class has been registered under a ``parametrizations`` attribute.
+# This ``parametrizations`` attribute is an ``nn.ModuleList``, and it can be accessed as such
+print(layer.parametrizations.weight[0])
+
+###############################################################################
+# Note that each element in the `ModuleList` is itself a list, and we have to select the first
+# element of this list. It will be clear later the reason for this, when we see how to contactenate
+# paramtrizations.
+#
+# Something that we may notice is that, if we print the parameters, we see that the
+# parameter ``weight`` has been moved
+print(dict(layer.named_parameters()))
+
+###############################################################################
+# It now sits under ``layer.parametrizations.weight.original``
+print(layer.parametrizations.weight.original)
+
+###############################################################################
+# Besides these two small differences, the parametrization is doing exactly the same
+# as our manual implementation
+symmetric = Symmetric()
+assert torch.allclose(layer.weight, symmetric(layer.parametrizations.weight.original))
+
+###############################################################################
+# Parametrizations are first-class citizens
+# -----------------------------------------
+# Since ``layer.parametrizations`` is an `nn.ModuleList`, it means that the parametrizations
+# are properly registered as submodules of the original module. As such, the same rules
+# for registering parameters in a module apply to register a parametrization.
+# For example, if a parametrization has parameters, these will be moved from CPU
+# to CUDA when calling ``model = model.cuda()``.
+#
+# Caching the value of a parametrization
+# --------------------------------------
+# Parametrizations come with an inbuilt caching system via the context manager ``P.cached()``
+class NoisyParametrization(nn.Module):
+    def forward(self, X):
+        print("Computing the Parametrization")
+        return X
+
+layer = nn.Linear(2, 3)
+P.register_parametrization(layer, "weight", NoisyParametrization())
+print("Here, layer.weight is recomputed every time we call it")
+foo = layer.weight + layer.weight.T
+bar = layer.weight.sum()
+with P.cached():
+    print("Here, it is computed just the first time layer.weight is called")
+    foo = layer.weight + layer.weight.T
+    bar = layer.weight.sum()
+
+###############################################################################
+# Composing Parametrizations
```

Comment:
We use many names for the same thing here, we should pick one I think between composing, chain or concatenate.

Reply:
Ooops. Done.